_posts/2021-08-18-Day-8-Post.md

```diff
@@ -0,0 +1,123 @@
+---
+layout: post
+title: "If I Could Call An Argument, Do You Think Const Would Pass Me By?"
+author: Kevin Fong
+---
+## Argument Passing by Copy, Reference, Pointer
+
+There are many different ways to call arguments into a function in C++. Unless specified, arguments are passed into function in C++ by copying the contents. Because the copy is stored in a new address, changes that are made to the data within the function are not reflected in the original data outside of the function. In the case we want to mutate the data we are calling into a function, we will need to refer to the data with `&` in the signature of the function. Because a reference simply refers to a variable, a reference can be made to modify the variable's contents through the variable. Calling the reference should be the default way to pass in data that will be changed by the function. 
+```cpp
+#include <iostream>
+void convertFtoC(double & temperature)
+{
+    temperature= (temperature - 32)*(5.0/9.0);
+}
+
+int main(void)
+{
+    double temperature = 212.0;
+    std::cout<<"Original temperature is " <<temperature<<"F."<<std::endl;
+    convertFtoC(temperature);
+    std::cout<<"New temperature is " <<temperature<<"C."<<std::endl;
+    return 0;
+}
+```
+It is also possible to passing data by point. By passing the data with `*` in the signature, the address of the variable will be passed into the function. Passing by pointer is sometimes used to pass around arrays allocated with `new`, after which elements are accessed with `*`. However, this is awkward for other uses, where passing by references are preferred.
+
+## Const referencing
+
+A `const` reference can be passed into a function to make an imutable variable. Once a const variable is passed into a function, it cannot be changed within the function. However it is important to maintian *const correctness* when use const references. A `const` reference can accept both `const` and non-`const` data, but non-`const` reference can only take non-`const` objects. Therefore, a non-`const` reference cannot be made of a `const` object, you can't try to refer and change a `const` object. It's generally recommending to make referene arguments const if you do not have any intention of modifying the data, and is importing for maintaining the integrity of large data structures, like vectors and strings. It is much safer to write functions that do not mutate their input parameters, and to avoid global parameters.
```

Comment:
```suggestion
A `const` reference can be passed into a function to make an immutable variable. Once a const variable is passed into a function, it cannot be changed within the function. However it is important to maintain *const correctness* when use const references. A `const` reference can accept both `const` and non-`const` data, but non-`const` reference can only take non-`const` objects. Therefore, a non-`const` reference cannot be made of a `const` object, you can't try to refer and change a `const` object. It's generally recommending to make reference arguments const if you do not have any intention of modifying the data, and is importing for maintaining the integrity of large data structures, like vectors and strings. It is much safer to write functions that do not mutate their input parameters, and to avoid global parameters.
```